src/trig.c

```diff
@@ -97,3 +97,102 @@ void twin_sincos(twin_angle_t a, twin_fixed_t *sin, twin_fixed_t *cos)
         *cos = cos_val;
     }
 }
+
+static const twin_angle_t atan_table[] = {
+    0x0200, /* arctan(2^0)  = 45° -> 512     */
+    0x0130, /* arctan(2^-1) = 26.565° -> 303 */
+    0x009B, /* arctan(2^-2) = 14.036° -> 155 */
+    0x004F, /* arctan(2^-3) = 7.125° -> 79   */
+    0x0027, /* arctan(2^-4) = 3.576° -> 39   */
+    0x0014, /* arctan(2^-5) = 1.790° -> 20   */
+    0x000A, /* arctan(2^-6) = 0.895° -> 10   */
+    0x0005, /* arctan(2^-7) = 0.448° -> 5    */
+    0x0002, /* arctan(2^-8) = 0.224° -> 2    */
+    0x0001, /* arctan(2^-9) = 0.112° -> 1    */
+    0x0001, /* arctan(2^-10) = 0.056° -> 1   */
+    0x0000, /* arctan(2^-11) = 0.028° -> 0   */
+};
+
+static twin_angle_t twin_atan2_first_quadrant(twin_fixed_t y, twin_fixed_t x)
+{
+    if (x == 0 && y == 0)
+        return 0;
+    if (x == 0)
+        return TWIN_ANGLE_90;
+    if (y == 0)
+        return TWIN_ANGLE_0;
+    twin_fixed_t current_x = x;
+    twin_fixed_t current_y = y;
+    twin_angle_t angle = 0;
+    /* CORDIC iteration */
+    for (int i = 0; i < 12; i++) {
+        twin_fixed_t temp_x, temp_y;
+        if (current_y > 0) {
+            temp_x = current_x + (current_y >> i);
+            temp_y = current_y - (current_x >> i);
+            angle += atan_table[i];
+        } else {
+            temp_x = current_x - (current_y >> i);
+            temp_y = current_y + (current_x >> i);
+            angle -= atan_table[i];
+        }
+        current_x = temp_x;
+        current_y = temp_y;
+    }
+    return angle;
+}
+
+twin_angle_t twin_atan2(twin_fixed_t y, twin_fixed_t x)
+{
+    if (x == 0 && y == 0)
+        return TWIN_ANGLE_0;
+    if (x == 0)
+        return (y > 0) ? TWIN_ANGLE_90 : TWIN_ANGLE_270;
+    if (y == 0)
+        return (x > 0) ? TWIN_ANGLE_0 : TWIN_ANGLE_180;
+    int quadrant;
+    twin_fixed_t abs_x = x;
+    twin_fixed_t abs_y = y;
+    if (x >= 0 && y >= 0) {
```

Comment:
Perhaps it can be modified to a branch-free method for determining the quadrant of the angle.

The return values seem to follow a certain pattern, which can be expressed in the following algebraic form as  $\text{result} = 2048 * \text{m} + \text{sign} * \text{angle}$.

Reply:
If you still have no idea to implement, ChatGPT is a good helper.

Reply:
@jouae Thanks for the detailed explanation : )